### ISSUE TITLE
Non-ASCII symbols support

### ISSUE THREAD

**Issue by rkononov** [yxtay/char-rnn-text-generation#1]
I'm trying to use this project to generate text with non-ascii symbols (cyrilic) and keep getting spaces/tabs/commas but no text, is this related to string.printable which part i should modify to enable support of non-ascii symbols?

Thanks!

Output

python tf_model.py train --text-path Vojna_i_mir._Kniga_1.txt --checkpoint-path /home/norn/src/char-rnn-text-generation/tf_checkpoint/chk 
2017-12-16 15:10:56,234 - __main__ - INFO - corpus length: 1433026.
2017-12-16 15:10:56,235 - __main__ - INFO - building model: {'clip_norm': 5.0, 'batch_size': 64, 'num_layers': 2, 'vocab_size': 98, 'rnn_size': 128, 'p_keep': 1.0, 'learning_rate': 0.001, 'embedding_size': 32}.
2017-12-16 15:10:57.231924: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA
2017-12-16 15:10:57,523 - __main__ - INFO - model saved: /home/norn/src/char-rnn-text-generation/tf_checkpoint/chk.
2017-12-16 15:10:57,739 - __main__ - INFO - tensorboard set up.
2017-12-16 15:10:57,740 - __main__ - INFO - building model: {'clip_norm': 5.0, 'batch_size': 1, 'num_layers': 2, 'vocab_size': 98, 'rnn_size': 128, 'p_keep': 1.0, 'learning_rate': 0.001, 'embedding_size': 32}.
2017-12-16 15:10:57,996 - __main__ - INFO - inference model loaded: /home/norn/src/char-rnn-text-generation/tf_checkpoint/chk.
2017-12-16 15:10:58,530 - __main__ - INFO - start of training.
epoch 1/32:   0%|                                                                                                             | 0/349 [00:00<?, ?it/s]2017-12-16 15:10:58,532 - utils - INFO - number of batches: 349.
2017-12-16 15:10:58,532 - utils - INFO - effective text length: 1429504.
2017-12-16 15:10:58,532 - utils - INFO - x shape: (64, 22336).
2017-12-16 15:10:58,532 - utils - INFO - y shape: (64, 22336).
epoch 1/32: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 349/349 [02:53<00:00,  1.97it/s]
2017-12-16 15:13:51,930 - __main__ - INFO - epoch: 1, duration: 173s, loss: 0.93437.
2017-12-16 15:13:52,082 - __main__ - INFO - model saved: /home/norn/src/char-rnn-text-generation/tf_checkpoint/chk.
2017-12-16 15:13:52,114 - __main__ - INFO - generating 512 characters from top 10 choices.
2017-12-16 15:13:52,115 - __main__ - INFO - generating with seed: "а помощь к брату, кто бы он ни б".
2017-12-16 15:13:53,027 - __main__ - INFO - generated text: 
а помощь к брату, кто бы он ни б     ,      , ,   
  ,               ,-
        ,          ,         .  
,   .        ,       

epoch 2/32:

**Comment by yxtay** [comment]
You are right. Currently the dictionary of characters is fixed regardless of text input and only accepts characters `string.printable`, which is ASCII only. 

You can overcome this by generating the `id2char` and `char2id` dictionary mapping based on the text input and correct the code where `encode_text()` or `ID2CHAR` is used. `encode_text()` accepts a second argument of the `char2id` mapping. 

Creating dictionary based on data input can be done with slight modification of the `create_dictionary()` code in `utils.py` as follows:
```
chars = sorted(set(text))
```

In order to generate text from the trained model, the dictionary created will have to be serialised in some form (JSON perhaps) and subsequently loaded.

Hope this helps.

**Comment by rkononov** [comment]
Thank you, yeah I figured out how to expand dictionary.

BTW side question what you think about generating not char by char but word by word?

**Comment by yxtay** [comment]
Generating word by word requires a bit more text preprocessing. Especially problematic would be the handling of punctuation and tokenisation, although a naive approach may work.

Another matter to consider would be the large size of vocabulary due to rare and misspelled words. Some form of vocabulary trimming would be necessary. The large vocabulary size may also be problematic during training due to the large number of parameters to update for the embedding and dense layers.
